Allow jsonapi-rspec to match Symbol or String based source documents

sa73917 opened the pull request:
## What is the current behavior?

The current version of the jsonapi-rspec matcher requires the source jsonapi document to be a string keyed hash and requires all hash parameters to use the same key style as the source jsonapi document.  

## What is the new behavior?

All matchers now support:
- a string or symbol keyed jsonapi document
- (where applicable) a string or symbol keyed parameter hash (or array of hashes)

## Checklist

Please make sure the following requirements are complete:

- [X] Tests for the changes have been added (for bug fixes / features)
All Matchers have a spec that tests the matcher against Symbol and String keyed jsonapi documents.  The Relationships matcher did not have an spec so I have created one based on the sample matchers in the README and the code.
Where applicable the tests include Symbol and String keyed hash parameters

- [X] Docs have been reviewed and added / updated if needed (for bug fixes / features)
README.md document has been updated to provide an example spec helper, and examples of String and Symbol based keys.

- [X] All automated checks pass (CI/CD)
rubocop and rspec tests pass.  
NB1. I have added the new rubocop tests and a few cop defaults to rubocop.yml in 633b06. 
NB2. I have added spec_helper into .rspec to save including it in all specs in 33eb64.

stas commented:
@sa73917 thank you for the contribution and the nice PR!!!

Before we proceed with the full review, I'd like to raise some questions and try to understand your approach a bit on a higher level...

Particularly, I believe you took the _ActiveSupport_ approach here, and ended up monkey-patching tested data (based on the `IndifferentAccess` module, we're overwriting core `Hash#dig` and other accessor methods). I generally don't find this a good idea and try to encourage everyone to stay away from such technics.

Would you reconsider the approach and instead use the core [`JSON#parse`](https://ruby-doc.org/stdlib-2.6.3/libdoc/json/rdoc/JSON.html#parse-method) method to take care of the symbol/string names instead?

Consider this...
```ruby
module JSONAPI
  module RSpec
    mattr_accessor :use_indifferent_access

    def as_indifferent_hash(doc)
      return doc unless self.use_indifferent_access
      return doc.with_indifferent_access if doc.respond_to?(:with_indifferent_access) && self.use_indifferent_access
      JSON.parse(JSON.generate(doc))
    end

    module Meta
      ::RSpec::Matchers.define :have_meta do |val|
        match do |actual|
          actual = JSONAPI::RSpec.as_indifferent_hash(actual)
          actual.key?('meta') &&
            (!val || actual['meta'] == JSONAPI::RSpec.as_indifferent_hash(val))
        end
      end
    end
  end
end
```

An in a similar manner just convert the names to string if the indifferent access is set.

Another important thing to mention is that I believe this should be an absolutely optional feature. JSONAPI spec clearly communicates that the format is JSON, it's just the Ruby that blurries the lines between how JSON can be parsed/handled. As in, a parsed JSON response would never have symbols or mixed key names in it...

Let me know what you think and thanks again for the contribution!

sa73917 commented:
Hi @stas , thanks for the quick reply :).  I'm more than happy to consider any approach as I'm pretty new to ruby (and this being a first PR here, open to feedback).. 

Slotting in `JSON.parse(JSON.generate(doc))` and using key'd hash comparisons slotted in pretty easily - definitely simpler than how I was doing it.

Reading through the code you've provided above to add a toggle for indifferent access (assuming I'm reading things right), you're relying on various rails methods/variables (`mattr_accessor`, `with_indifferent_access`).  Was this the intent? I was trying to achieve the 'indifferent' result without requiring any additional gems to save on bloating this one.  Of couse there's every possibility there's a way that I don't know of to use these rails features without bringing in all of ActiveSupport - see previous comment about being new to all this! :-)

I agree with you that the jsonapi payload should be JSON rather than a Symbol key'd hash but symbolizing the payload makes other (non jsonapi-rspec) tests far more readable and having two versions of the payload, one for jsonapi-rspec tests and one for all the others seems wonky - hence this PR.

Anyway, I was playing around with your example this morning (though with a manually created setter/getter for the toggle in the module) and as written Meta was just returning `undefined method 'as_indifferent_hash' for JSONAPI::RSpec:Module`.  Probably just me doing something stupid but to work around that I made it a static method and it at least ran, but then was getting some fairly random results around what the `use_indifferent_access` toggle was set to when I got into the matcher (even if I set it inside the `it` directly before the `is expected`!?).  I wasn't getting very far so  decided to try a chain to toggle indifferent on - just to see how that would work.  eg:

```ruby
      ::RSpec::Matchers.define :have_type do |expected|
        match do |actual|
          actual = JSONAPI::RSpec.as_indifferent_hash(actual) if @use_indifferent_access

          (actual || {})['type'] == expected
        end

        chain :with_indifferent_access do
          @use_indifferent_access = true
        end
      end
```
which makes it obvious in the test

`     it { is_expected.to have_type('bar').with_indifferent_access }`

and makes the output pretty clear:

```ruby
JSONAPI::RSpec::Type
  #have_type
    with empty document
      is expected not to have type "bar"
    with stringifyed doc
      is expected to have type "bar" with indifferent access
      is expected not to have type "foo" with indifferent access
      without indifferent access
        is expected to have type "bar"
    with symbolized doc
      is expected not to have type "foo" with indifferent access
      is expected to have type "bar" with indifferent access
      without indifferent access
        is expected not to have type "bar"
```
But made me wonder whether adding `.with_indifferent_access` to all the tests would be a pain in the proverbial if someone's doing a lot of tests on the payload.  Anyway - I'm supposed to be doing my day job so I'll have another look later and see if I can get your idea working.

stas commented:
> Reading through the code you've provided above to add a toggle for indifferent access (assuming I'm reading things right), you're relying on various rails methods/variables (`mattr_accessor`, `with_indifferent_access`). Was this the intent? I was trying to achieve the 'indifferent' result without requiring any additional gems to save on bloating this one. Of couse there's every possibility there's a way that I don't know of to use these rails features without bringing in all of ActiveSupport - see previous comment about being new to all this! :-)

Yup, you're right, in fact, `attr_accessor` is more than enough for the configuration, alternatively we could use the [`RSpec#add_setting`](https://relishapp.com/rspec/rspec-core/v/3-8/docs/configuration/custom-settings), either way works. As for `with_indifferent_access`, I just assume a lot of folks might be using this gem along with the rails, so, no extra legwork is needed if the method is available...

> Anyway, I was playing around with your example this morning (though with a manually created setter/getter for the toggle in the module) and as written Meta was just returning `undefined method 'as_indifferent_hash' for JSONAPI::RSpec:Module`.

Apologies for that, I was in a rush, please look at the code I pasted as more of a pseudocode. It's definitely not working :see_no_evil: 

> ` it { is_expected.to have_type('bar').with_indifferent_access }`

Indeed, this is actually a nice idea, but I believe it would be a lot to type for some folks, so a more global option would be desired...


Thanks again for working on this! :bow:

sa73917 commented:
I decided rather than add a getter and setter for a class variable in the module, that using a custom  RSpec configuration variable would make more sense (thanks for pointing that out).   By default jsonapi-rspec will expect actual JSON documents, but setting `RSpec.configuration.allow_symbolized_jsonapi = true` will allow it to support Symbol keyed hashes.

I've removed all the monkey patch logic so the PR now works off the "as_different_hash" method you posted initially.  Happy to tweak anything else if you see something in your review.. 

Couple of side notes:
1. I've made all the parameter values (`.with_data`, jsonapi value, meta value), support symbol's / symbol keyed hashes as well as strings - which I think is what people would expect?
2. You'll see in the specs I'm toggling the allow_symbolized_jsonapi setting to true/false `before(:each)`.  In normal usage you'd be running it in one mode or the other, so could just set it once after the describe (as I've shown in the readme).  For the tests here however, as they are running in both modes (and Rspec seems to randomise the run order of them), it needs to be set before each or you get random failures :-).

sa73917 commented:
@stas - I think it's in a pretty good state to review now.  

I've removed the monkey-patching and used the JSON parsing logic (though I do wonder if this will have a performance impact - hopefully not given most of these payloads are unlikely to be huge?)

I've refactored all the rspec tests so they all loop through:
- the various source document types (empty, HashWithIndifferentAccess, symbolized hash and JSON)
- the two settings for the allow_symbolized_jsonapi configuration setting
- where applicable, the various formats the parameter value(s) can be in (symbolized, stringifyed)
This does make them a little more complicated but at least they should cover all the different combinations now (_I think_).  

I've made the allow_symbolized_jsonapi setting impact how the source document is handled but it doesn't affect the parameter values.  ie. { version: '1.0' } and { "version" => "1.0" } are both accepted by the jsonapi_object matcher regardless of the setting.

Let me know if there's anything more you think I should change.

sa73917 commented on README.md:
```diff
@@ -29,42 +25,69 @@ RSpec.configure do |config|
 ## Usage and documentation
 
-Available matchers:
+### Create a Helper to Convert the Response (Optional)
```
I added this in as the first time I went to use the gem, it was very unclear what `document` was in the `expect(document).to have_meta` examples.  You're right it's opinionated so happy to remove it but but I do think the readme should explain what `document` is.  Ruby being duck typed, it could be anything :-)

Something like 
```ruby
# Example usage
let(:document) { JSON.parse(response.body, symbolize_names:  false) }
```
above the available matchers might help?

sa73917 commented on README.md:
````diff
@@ -29,42 +25,69 @@ RSpec.configure do |config|
+RSpec.describe 'Authorization Requests', type:  :request  do
+	RSpec.configuration.allow_symbolized_jsonapi = true
+	...
+```
````
Happy to adjust the description of the custom setting, but I think

 > The configuration above allows using either string or symbol names with the RSpec matchers.

isn't very clear as it doesn't specify whether we're talking about the JSONAPI document we're testing, or the parameter values we're passing in.  I've adjusted the README to simplify the indifferent matching section.

sa73917 commented on lib/jsonapi/rspec/attributes.rb:
```diff
@@ -1,26 +1,27 @@
+require 'pry'
```
My bad - it'd been a long day :)

sa73917 commented on lib/jsonapi/rspec/attributes.rb:
```diff
@@ -1,26 +1,27 @@
 
         chain :with_value do |val|
-          @val_set = true
+          @with_value = true
```
In hindsight - definitely should be in its own "variable naming consistency" type PR.  I'll revert these but I think it would be good to standardise on a naming format for the chain booleans.  One is the method name and the others are [something]_set. 

@val_set isn't very clear (

sa73917 commented on .rubocop.yml:
```diff
@@ -1,8 +1,58 @@
+  Max: 15
+
+Layout/LineLength:
+  Max: 250
```
Apologies - copied over some of my defaults that shouldn't have made it in :)

sa73917 commented:
I've taken note of the various notes above (thanks for all the input - sorry you're the poor person on the end of a first PR for me)..   I've split out all the code around supporting symbol parameters into a distinct PR and made the changes you've suggested there.  Will work with you on that PR first and hopefully get it merged in - then rebase this one and see what's left.  Will know for future never put more than one "change" into a single PR! Sorry again..

sa73917 commented:
@stas - finally had a chance to go through and refactor everything to split out the two changes that were in this PR.

PR #19 has the code to allow jsonapi-rspec to accept Symbol/String Parameters, and this one has the code to allow jsonapi-rspec to match against Symbol/String source documents.  (obviously a rebase in one of them's future)

Hopefully a little easier to review - thanks for your patience :-)

sa73917 commented:
@stas - just checking in to see if there's anything you're waiting on me for?  Would love to finalise this PR so I can rebase PR #19 and get that one ready for review :)

stas commented on spec/jsonapi/attributes_spec.rb:
```diff
@@ -1,25 +1,41 @@
+
+      it { is_expected.to have_attribute(:one) }
+      it { is_expected.not_to have_attribute(:five) }
+    end
```
@sa73917 please try to avoid such loops in the tests, these are harder to read and maintain. You did a great job DRY-ing up the tests so all of these should be one-liners, but it helps a lot extending and maintaining the tests long-term.

Please maintain the explicit calls to `expect` instead of `is_expected`.

stas replied:
Also, as I mentioned in the previous PR, we don't need to run both use-cases for every matcher. If it makes sense, let's add a unit-test to the `JSONAPI::RSpec.as_indifferent_hash` and since these all are calling it, it's more than enough to provide just one or two examples using string-based keys. The idea is to isolate, test and re-use component/parts of the code, and not carry over the tests for one component over and over. Let me know if this makes sense.

sa73917 commented:
@stas I've run through and refactored all the specs as requested.  Makes sense to have test of as_indifferent_hash and keep the noise out of the other tests so I've put together one that tests the various combinations.  All of the existing tests are now in the same format, all include a present/not present test and where applicable include symbol and string parameters.
